sqlmesh/core/engine_adapter/spark.py

```diff
@@ -59,6 +59,46 @@ def insert_append(
         else:
             super().insert_append(table_name, query_or_df, columns_to_types, contains_json)
 
+    def merge(
+        self,
+        target_table: TableName,
+        source_table: QueryOrDF,
+        column_names: t.Iterable[str],
+        unique_key: t.Iterable[str],
+    ) -> None:
+        if isinstance(source_table, PySparkDataFrame):
+            self._merge_pyspark_df(target_table, source_table, column_names, unique_key)
+        else:
+            super().merge(target_table, source_table, column_names, unique_key)
+
+    def _merge_pyspark_df(
+        self,
+        target_table: TableName,
+        source_df: PySparkDataFrame,
+        column_names: t.Iterable[str],
+        unique_key: t.Iterable[str],
+    ) -> None:
+        """
+        Merge using DataFrames is only supported for Delta Tables using the Delta Table API.
```

Comment:
Shouldn't this go into the engine adapter specific to Databricks then?

Reply:
So the answer was no because you can use Delta lake outside of Databricks. @tobymao has a great suggestion though of using temp view to use SQL instead of Dataframe operations which will allow this to work for Iceberg too. Great idea @tobymao!